server/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java

```diff
@@ -159,7 +159,10 @@ private class AsyncShardStoresInfoFetches {
         void start() {
             try {
                 for (Tuple<ShardId, String> shard : shards) {
-                    new InternalAsyncFetch(logger, "shard_stores", shard.v1(), shard.v2()).fetchData(nodes, Collections.emptySet());
+                    new InternalAsyncFetch(logger, "shard_stores", shard.v1(), shard.v2(), this.nodes.getSize()).fetchData(
```

Comment:
We only need allocate enough to hold all the data nodes so `routingNodes.size()` is better here:

```suggestion
                    new InternalAsyncFetch(logger, "shard_stores", shard.v1(), shard.v2(), routingNodes.size()).fetchData(
```